### DISCUSSION TITLE
Trim Failed - no clue about

### DISCUSSION THREAD

**Discussion by gasilva** [JSBSim-Team/jsbsim#490] [Q&A]
HI,

```shell
Full Trim
  Sorry, qdot doesn't appear to be trimmable

  Trim failed
  Trim Results: 
       Angle of Attack:  10.16  wdot: -2.52e+00 Tolerance: 1e-03  Failed
              Throttle:   0.81  udot: -2.59e+00 Tolerance: 1e-03  Failed
            Pitch Trim:   0.00  qdot: -1.46e+00 Tolerance: 1e-04  Failed
            Roll Angle:   0.00  vdot: -1.04e-17 Tolerance: 1e-03  Passed
              Ailerons:   0.00  pdot:  6.28e-10 Tolerance: 1e-04  Passed
                Rudder:   0.00  rdot: -8.18e-10 Tolerance: 1e-04  Passed
Traceback (most recent call last):
  File "/Users/gasilva/Documents/GitHub/jsbsimNew/examples/python/RunScriptLongFlight.py", line 59, in <module>
    while fdm.run():
  File "jsbsim.pyx", line 1010, in jsbsim.FGFDMExec.run
RuntimeError: Trim Failed
```

I am implementing a Global5000 aircraft with DATCOM aerodynamics. Everything is running OK. But when I run trim-cruise_ap.xml I got the error above.

How can I debug it and know that is happening? Is there any clue to solve this problem?

Regards,

**Comment by seanmcleod** [comment]
What are the conditions you're trying to trim at, i.e. airspeed, altitude etc.?

For example maybe the airspeed you're using is too low, such that given the the alpha range the trim routines are using it isn't possible to generate enough lift for the given weight etc. Or maybe it can, but there is too much drag at that alpha such that even 100% throttle isn't enough.

**Comment by gasilva** [comment]
Basically it was a wrong forces balance model from DATCOM Pro.
They implemented a totally different force balance model and a table for
drag instead of using Drag-basic and Drag-induced.

I migrated all DATCOM data to Aeromatic format and it worked.
Thanks.

PS: I can put the wrong file to your comments if you want.

**Comment by seanmcleod** [comment]
There is no reason that a table for drag shouldn't work. JSBSim doesn't care how the forces and moments are defined.

So I'd be interested to see what the original DATCOM data looked like.

**Comment by agodemar** [comment]
@gasilva Interesting work!
Your model of Global5000 and your tests on trimming, are they  going to be available or explained somewhere in our Wiki?

**Reply by gasilva** [reply]
Maybe yes. But I need to know how to setup a blog like @seanmcleod did.

**Reply by seanmcleod** [reply]
Take a look at - https://pages.github.com/ or as I mentioned previously you can also just add jupyter notebooks to a repo and reference them.

**Reply by agodemar** [reply]
@gasilva @seanmcleod @bcoconni 
I think it will be the case to host a *JSBSim Blog*. The supporting technology is [GitHub Pages](https://pages.github.com/), much as we do for our [JSBSim Online Reference Manual](https://jsbsim-team.github.io/jsbsim-reference-manual/).
This time I'll try out [Docusaurus](https://docusaurus.io/), based on [Node.js](https://nodejs.org/) and [Yarn](https://classic.yarnpkg.com/en/).
The sources of the blog will be public, contributions by users will be welcomed. All contributions will be subject to the usual PR workflow. 
What do you think?
(I'll move this post to a separate discussion)

**Comment by gasilva** [comment]
global5000_aero.xml
file created by DATCOM Pro

[global5000_aero.xml.zip](https://github.com/JSBSim-Team/jsbsim/files/6911379/global5000_aero.xml.zip)


Aircraft file that used global5000_aero.xml

[global5000DAT.xml.zip](https://github.com/JSBSim-Team/jsbsim/files/6911376/global5000DAT.xml.zip)

**Comment by gasilva** [comment]
Global5000 file with DATCOM information in the format of Aeromatic.
This actually worked.

**Reply by gasilva** [reply]
[global5000AeroDAT.xml.zip](https://github.com/JSBSim-Team/jsbsim/files/6911374/global5000AeroDAT.xml.zip)

**Reply by gasilva** [reply]
The script is not trimming is this one.

[trim-cruise_ap.xml.zip](https://github.com/JSBSim-Team/jsbsim/files/6911382/trim-cruise_ap.xml.zip)

**Reply by gasilva** [reply]
you will need this also to run it
[airborne.xml.zip](https://github.com/JSBSim-Team/jsbsim/files/6911383/airborne.xml.zip)

**Reply by gasilva** [reply]
Consumption with a Mach step in autopilot

![Figure_fuel](https://user-images.githubusercontent.com/477692/127744797-ff307c28-6cf0-4d06-bda4-28ff81222d5a.png)
![Figure_Mach](https://user-images.githubusercontent.com/477692/127744799-6cccc0f5-8468-4404-a1c9-532b93c06bb4.png)

Consumption with a Mach ramp in autopilot
![Figure_Fuel2](https://user-images.githubusercontent.com/477692/127744886-07499180-7ce5-45c6-b28c-ff5853d0d145.png)
![Figure_Mach2](https://user-images.githubusercontent.com/477692/127744882-0f38bd80-a798-420e-bc88-ff0257096532.png)

**Comment by seanmcleod** [comment]
@gasilva have you compared the drag data from DATCOM versus from Aeromatic? For example plot Cd vs AoA for both.

**Reply by gasilva** [reply]
![Picture1](https://user-images.githubusercontent.com/477692/127871384-d1ea24e3-6deb-4aa0-af0e-23ad14d0fecb.png)
![Picture2](https://user-images.githubusercontent.com/477692/127871392-1a92f6e1-7f14-476c-b4fb-0d91b50c4c5d.png)

**Reply by gasilva** [reply]
DATCOM table was generated by the Software
JSBSim is the Aeromatic
DATCOM Analytic is the $C_D=C_{D0}+k \cdot C_L^2$

**Reply by gasilva** [reply]
![Picture3](https://user-images.githubusercontent.com/477692/127874805-a6ab853a-e112-4bf5-95e2-171f42b1994e.png)
![Picture4](https://user-images.githubusercontent.com/477692/127874816-e35384e1-fb08-48d5-ae8f-2127d7df7a61.png)

Just another fitting fro DATCOM data

**Reply by gasilva** [reply]
the Drag Analytic done with fitting of DATCOM table

```XML
    <function name="aero/force/Drag_basic">
       <description>Drag at zero lift</description>
       <product>
          <property>aero/qbar-psf</property>
          <property>metrics/Sw-sqft</property>
          <table>
            <independentVar lookup="row">aero/alpha-rad</independentVar>
            <tableData>
             -1.57    1.500
             -0.26    0.013474
              0.00    0.011474
              0.26    0.013474
              1.57    1.500
            </tableData>
          </table>
       </product>
    </function>

    <function name="aero/force/Drag_induced">
       <description>Induced drag</description>
         <product>
           <property>aero/qbar-psf</property>
           <property>metrics/Sw-sqft</property>
           <property>aero/cl-squared</property>
           <value>0.0583</value>
         </product>
    </function>
```

**Reply by gasilva** [reply]
> ![Picture3](https://user-images.githubusercontent.com/477692/127874805-a6ab853a-e112-4bf5-95e2-171f42b1994e.png)
> ![Picture4](https://user-images.githubusercontent.com/477692/127874816-e35384e1-fb08-48d5-ae8f-2127d7df7a61.png)
> 
> Just another fitting fro DATCOM data

The grey line from DATCOM transformed to Drag basic + induced works! The problem is the table.

**Comment by seanmcleod** [comment]
Out of interest what did you use to plot the graphs? Is it Matplotlib with a particular style?

So comparing the DATCOM (orange) with the Aeromatic (blue) I can see from about +9 AoA and larger you could run into a case where the trim works using the Aeromatic drag data but fails with the DATCOM data because there isn't enough thrust available.

At what sorts of speed/AoA did the trim start failing for the DATCOM data?

**Reply by gasilva** [reply]
Sorry, it is a plain Excel plots. I did not have time to make txt files and code. For sure, Python is better but...

The problem with DATCOM are not the Drag values. I tried to implement a Drag table as it is in DATCOM aero.xml but it did not worked. When I used Drag Basic and Induced Drag it worked.

The CD_0 and k worked well. But the table not.

```XML
 <axis name="DRAG">

    <function name="aero/force/Drag_basic">
       <description>Drag at zero lift</description>
       <product>
          <property>aero/qbar-psf</property>
          <property>metrics/Sw-sqft</property>
          <table>
            <independentVar lookup="row">aero/alpha-rad</independentVar>
            <tableData>
             -1.57    1.500
             -0.26    0.013474
              0.00    0.011474
              0.26    0.013474
              1.57    1.500
            </tableData>
          </table>
       </product>
    </function>

    <function name="aero/force/Drag_induced">
       <description>Induced drag</description>
         <product>
           <property>aero/qbar-psf</property>
           <property>metrics/Sw-sqft</property>
           <property>aero/cl-squared</property>
           <value>0.0583</value>
         </product>
    </function>
```

the table is like that

```XML
	<function name="aero/coefficient/CD_wbh">
		<description>
			Drag coefficient due to wing, body, and horizontal tail
			Main contributor to Phugoid damping: greater Cd, better damping
		</description>
		<table>
			<independentVar lookup="row"> aero/alpha-deg </independentVar>
			<tableData>
				     -8.0000         .0350
				     -6.0000         .0255
				     -4.0000         .0183
				     -2.0000         .0138
				       .0000         .0121
				      2.0000         .0139
				      4.0000         .0186
				      6.0000         .0265
				      8.0000         .0379
				     10.0000         .0526
				     12.0000         .0706
				     14.0000         .0922
				     16.0000         .1171
				     18.0000         .1453
			</tableData>
		</table>
	</function>
```

and the force balance is like that


```XML
	<axis name="DRAG" frame="STABILITY">

		<function name="aero/force/drag-lbs">
			<description>
				Drag force = ( CD_wbh + dCD_Min_Flap_1 + dCD_Flap + CD_Gear + dCD_Elev + CD_induced)
				             * q_bar * wing area
			</description>
			<product>
				<sum>
					<property> aero/coefficient/CD_wbh </property>
					<property> aero/coefficient/dCD_Min_Flap_1 </property>
					<property> aero/coefficient/dCD_Flap </property>
					<property> aero/coefficient/CD_Gear </property>
					<property> aero/coefficient/dCD_Elev </property>
				</sum>
				<property> aero/qbar-psf </property>
				<property> metrics/Sw-sqft </property>
			</product>
		</function>

	</axis>
```

**Comment by gasilva** [comment]
I am running the simulation in Mach=0.43 to 0.5.
The alpha is from 11 to 7 more or less.

**Comment by seanmcleod** [comment]
I don't see anything obviously wrong with the drag table. 

When you say it doesn't work, does this mean you can't get it to trim at all across the M0.43 to M0.5 range, or just a portion at say the lower end of the Mach range?

**Comment by gasilva** [comment]
the trim does not work in 0.4 to 0.52 range with the table and force
balance.

Em seg., 2 de ago. de 2021 às 13:15, Sean McLeod ***@***.***>
escreveu:

**Reply by gasilva** [reply]
maybe it worked in a previous version of JSBSim...

**Reply by gasilva** [reply]
it does not work even with force balance of JSBSim. If I define a force instead of coefficient.

**Comment by seanmcleod** [comment]
I'll try and take a look by running it and see if I can spot any issues.

In the meantime I'd also suggest that you contact the developer of DATCOM Pro to let him know.

**Reply by gasilva** [reply]
OK!

**Reply by gasilva** [reply]
Just did that and referred to this discussion.

**Reply by gasilva** [reply]
He know something like this happened in the past. He will inform soon.

**Comment by gasilva** [comment]
The error changed now when I changed the script.

  Full Trim

  ```shell
Trim successful
  Trim Results: 
       Angle of Attack:  12.01  wdot: -6.60e-04 Tolerance: 1e-03  Passed
              Throttle:   0.77  udot: -8.20e-04 Tolerance: 1e-03  Passed
            Pitch Trim:  -0.47  qdot:  2.80e-05 Tolerance: 1e-04  Passed
            Roll Angle:   0.00  vdot:  2.24e-17 Tolerance: 1e-03  Passed
              Ailerons:   0.00  pdot: -1.73e-15 Tolerance: 1e-04  Passed
                Rudder:   0.00  rdot:  8.55e-19 Tolerance: 1e-04  Passed

Repeating Notify (Event 2) executed at time: 0.008333
    position/h-agl-ft = 30000.000000
    ap/mach_setpoint = 0.430000
    fcs/throttle-pos-norm[1] = 0.767165
    fcs/throttle-cmd-norm[1] = 0.768530
    ap/throttle-cmd = -0.001562
    velocities/mach = 0.430000
    propulsion/engine[0]/n1 = 83.706485
    propulsion/engine[1]/n1 = 83.706485
    propulsion/engine[0]/thrust-lbs = 3711.430011
    propulsion/engine[1]/thrust-lbs = 3711.430011
    velocities/vc-kts = 157.641416
    aero/alpha-deg = 12.006342
    simulation/frame = 1.000000


Hold heading and altitude (Event 1) executed at time: 10.016663
    Cal. airspeed (kts):   = 157.636832
    Altitude (AGL, ft):    = 30000.373858
    Weight (lbs):          = 80101.397946


Repeating Notify (Event 2) executed at time: 1800.007613
    position/h-agl-ft = nan
    ap/mach_setpoint = 0.500000
    fcs/throttle-pos-norm[1] = nan
    fcs/throttle-cmd-norm[1] = 0.768530
    ap/throttle-cmd = nan
    velocities/mach = nan
    propulsion/engine[0]/n1 = 30.000000
    propulsion/engine[1]/n1 = 30.000000
    propulsion/engine[0]/thrust-lbs = nan
    propulsion/engine[1]/thrust-lbs = nan
    velocities/vc-kts = 0.000000
    aero/alpha-deg = 0.000000
    simulation/frame = 216001.000000
```

**Comment by seanmcleod** [comment]
So we can close the discussion?

**Comment by gasilva** [comment]
Actually not. The drag table does not work.
Only if transformed to Drag Basic (CD0) and Drag Induced (k).

I tried to debug but no success.

**Comment by seanmcleod** [comment]
I'm confused, yesterday I received a Github notification for this discussion in which you mentioned - "no issue with DATCOM JSBSim file".

> When I copied one over another...I do not know what was different. Simply as that.
But no issue with DATCOM JSBSim file. It was the trim-cruise_ap.xml (trim cruise with autopilot).

And another one.

> I do not know why it started working. Just copy the script that worked on top of the other script. But now I do not have any bug tracking.

**Comment by gasilva** [comment]
Yes false alarm!
I was happy at that time.
But I discovered that I was running the other model by accident.

**Comment by seanmcleod** [comment]
Have you had any further feedback from the DATCOM Pro developer?

**Comment by gasilva** [comment]
Not yet. I gave the false alarm to him also.

Enviado do meu iPhone
